### PR TITLE
feat(locations): (2/4) set up models to return locations in GET /movies

### PR DESCRIPTION
**Pull request by karankwatra**
**Changes to models to support returning locations in GET /movies**
**Why:** Now that we have the db set up to support locations we can now return those relations when we get a movie. 
**What:** Created the locations model. Modified movies model to relate locations. Modified movies controller to fetch related locations.
**Notes:** I had to rename the db table to `locations_movies` because this is how Hapi/bookshelf(?) resolves these relations by default.

### CODE REVIEW

File: test/models/location.test.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+
+const Location = require('../../lib/models/location');
+
+describe('location model', () => {
+
+  describe('serialize', () => {
+
+    it('includes all of the necessary fields', () => {
+      const location = Location.forge().serialize();
```
Comment (Lianathanoj) [inline test/models/location.test.js]: nit: use the `new` identifier instead of `.forge()`. See [this discussion](https://github.com/lob/lens-api/pull/212#discussion_r323993648).

File: test/models/movie.test.js [hunk 2]
```diff
@@ -1,6 +1,9 @@
 'use strict';
 
-const Movie = require('../../lib/models/movie');
+const Knex = require('../../lib/libraries/knex');
+
+const Movie    = require('../../lib/models/movie');
+const Location = require('../../lib/models/location');
```
Comment (Lianathanoj) [inline test/models/movie.test.js]: nit: alphabetize imports

File: test/plugins/features/movies/controller.test.js [hunk 3]
```diff
@@ -71,4 +73,29 @@ describe('movie controller', () => {
 
   });
 
+  describe('retrive movies with locations', () => {
```
Comment (Lianathanoj) [inline test/plugins/features/movies/controller.test.js]: nit: typo on "retrieve"

**Comment by Lianathanoj** [comment]
Could you also add a method in the `movies` controller to associate locations with movies as specified in the ticket? 

> add POST /movies/{id}/locations to add new locations to a movie

File: test/plugins/features/movies/controller.test.js [hunk 4]
```diff
@@ -71,4 +73,29 @@ describe('movie controller', () => {
...
+    beforeEach(async () => {
+      await Knex.raw('TRUNCATE movies CASCADE; TRUNCATE locations CASCADE; TRUNCATE locations_movies CASCADE;');
+
+      const movie = await new Movie().save(movie_payload);
+      const location = await new Location().save(location_payload);
+
+      await new Movie({ id: movie.id }).locations().attach(location);
+    });
+
+    it('retrives a movie with the related locations', async () => {
```
Comment (Lianathanoj) [inline test/plugins/features/movies/controller.test.js]: nit: another typo on `retrieve`

File: test/models/movie.test.js [hunk 5]
```diff
@@ -13,10 +16,29 @@ describe('movie model', () => {
...
+        'locations',
         'object'
       ]);
     });
 
+    it('serialzes with location', async () => {
+      const movie_payload = { name: 'Zodiac' };
+      const location_payload = { name: 'Bay Bridge' };
+
+      await Knex.raw('TRUNCATE movies CASCADE; TRUNCATE locations CASCADE; TRUNCATE locations_movies CASCADE;');
```
Comment (Lianathanoj) [inline test/models/movie.test.js]: Could you place this statement in a `beforeEach()`? We usually only delete or truncate from tables before each test case to maintain consistency rather than having such an operation within the scope of only one case.

File: test/models/movie.test.js [hunk 6]
```diff
@@ -13,10 +16,29 @@ describe('movie model', () => {
         'id',
         'title',
         'release_year',
+        'locations',
         'object'
       ]);
     });
 
+    it('serialzes with location', async () => {
+      const movie_payload = { name: 'Zodiac' };
```
Comment (Lianathanoj) [inline test/models/movie.test.js]: could you camelCase these JS variables? here and elsewhere

File: test/models/movie.test.js [hunk 7]
```diff
@@ -13,10 +16,29 @@ describe('movie model', () => {
...
+      const location_payload = { name: 'Bay Bridge' };
+
+      await Knex.raw('TRUNCATE movies CASCADE; TRUNCATE locations CASCADE; TRUNCATE locations_movies CASCADE;');
+
+      const movie = await new Movie().save(movie_payload);
+      const location = await new Location().save(location_payload);
+      const loc = await new Location({ id: location.id }).fetch();
```
Comment (Lianathanoj) [inline test/models/movie.test.js]: not entirely sure this part where we redefine `loc` is necessary (for serialization at least). I know that for certain controller tests, we usually create a resource then fetch it from the DB to ensure that there aren't any issues with inserting into the DB, but we can kinda separate and test that logic out in the controller instead.

File: test/models/movie.test.js [hunk 8]
```diff
@@ -13,10 +16,29 @@ describe('movie model', () => {
...
+
+      await new Movie({ id: movie.id }).locations().attach(loc);
+
+      const movie_fetch = await new Movie({ id: movie.id }).fetch({ withRelated: ['locations'] });
```
Comment (Lianathanoj) [inline test/models/movie.test.js]: Make sure to actually call `serialize` if we want to test it! Right now this test looks more like a controller test, so I think an easier way to test for serialization in this case would be to 
1) Create a movie model
2) Create a location model
3) Set the movie's location attribute to the location (e.g. `movie.locations = location). 
4) Call `movie.serialize()` 
5) Verify that the serialized model has the correct location.

but lmk what you think!

File: test/models/location.test.js (hunk 1 above)
Comment (karankwatra) [reply test/models/location.test.js]: changed here and in the movie test as well

File: test/plugins/features/movies/controller.test.js (hunk 4 above)
Comment (karankwatra) [reply test/plugins/features/movies/controller.test.js]: lol apparently i cant spell this word

File: test/models/movie.test.js (hunk 8 above)
Comment (karankwatra) [reply test/models/movie.test.js]: i changed this test a little bit. good call on not needing `loc`. got away with using the `movie` and `location` variables from before. i do have to call one `.fetch`. This is to get the related locations using `withRelated`. I then serialize and check the fields. Let me know if this is more what you were thinking.

**Comment by karankwatra** [comment]
> Could you also add a method in the `movies` controller to associate locations with movies as specified in the ticket?
> 
> > add POST /movies/{id}/locations to add new locations to a movie

i was planning on holding this off for the next pr (with its tests). i could include it here if you think i should.

**Comment by Lianathanoj** [comment]
Oh yeah you can hold off until the next PR then for the association implementation!

File: db/migrations/20190919123945_create_movies_locations.js [hunk 9]
```diff
@@ -1,13 +1,13 @@
 'use strict';
 
 exports.up = async (Knex) => {
-  await Knex.schema.createTable('movies_locations', (table) => {
+  await Knex.schema.createTable('locations_movies', (table) => {
```
Comment (kienan) [inline db/migrations/20190919123945_create_movies_locations.js]: Is this change necessary? Do we know why if so?

Comment (karankwatra) [reply db/migrations/20190919123945_create_movies_locations.js]: I think this was changed because by default, bookshelf defines the joining table name alphabetically when fetching. I believe there is a way to override this. Let me know if I should revert this to conform to the previous PR.

File: test/plugins/features/movies/index.test.js [hunk 10]
```diff
@@ -1,11 +1,17 @@
...
 
+const Knex = require('../../../../lib/libraries/knex');
+
 const Movies = require('../../../../lib/server');
 
 describe('movies integration', () => {
 
   describe('create', () => {
 
+    beforeEach(async () => {
```
Comment (Lianathanoj) [inline test/plugins/features/movies/index.test.js]: I was going to say you don't need this `beforeEach` block, but I realized something else: right now, your current integration test for the the `retrieve` describe block relies on the movie created from the `create` describe block, whereas we usually try to make sure one test does not rely on another. While this is okay due to the smaller scope of these tests, could you actually move up this `beforeEach` block to the outer `describe` block while also adding DB insertion for movies via the factories you made? 

Moving it up to the outer block will allow the `beforeEach` statement to be run before every test case within its scope, whereas the current behavior only allows it to run for the `create` describe block.

File: test/plugins/features/movies/controller.test.js [hunk 11]
```diff
@@ -4,10 +4,24 @@ const Knex = require('../../../../lib/libraries/knex');
...
+const testMovie2 = MovieFactory.build({ name: 'Dawn of the Planet of the Apes', release_year: 2014 });
+
+const testLocation = LocationFactory.build();
+const testLocMov   = LocationMovieFactory.build({ movie_id: testMovie1.id, location_id: testLocation.id });
+
 describe('movie controller', () => {
 
   describe('create', () => {
 
+    beforeEach(async () => {
```
Comment (Lianathanoj) [inline test/plugins/features/movies/controller.test.js]: similar comment: could you move this `beforeEach` block to the outer `describe` block? In this case, it would reduce the need to have multiple `beforeEach` blocks and copypasta. 

One thing to note though is that sometimes we don't want to have a `beforeEach` at the outermost describe block if we want to be more specific with how we want to run certain statements before given `describe` blocks

File: test/plugins/features/movies/controller.test.js [hunk 12]
```diff
@@ -38,35 +49,55 @@ describe('movie controller', () => {
...
-      expect(movies.models[0].get('release_year')).to.eql(payload2.release_year);
+      expect(movies.models[0].get('name')).to.eql(testMovie2.name);
+      expect(movies.models[0].get('release_year')).to.eql(testMovie2.release_year);
+    });
+
+  });
+
+  describe('retrieve movies with locations', () => {
+
+    beforeEach(async () => {
```
Comment (Lianathanoj) [inline test/plugins/features/movies/controller.test.js]: if you manage to move the `describe` block up in scope, this `beforeEach` would no longer be needed too. However, the outer block should probably have the functionality to `TRUNCATE` + insert the movies/locations/locations_movies resources too.

File: test/plugins/features/movies/controller.test.js (hunk 11 above)
Comment (karankwatra) [reply test/plugins/features/movies/controller.test.js]: copypasta... definitely using that one haha

**Comment by karankwatra** [comment]
> This looks super clean now, nice! I really like how you employed factories to not only make testing easier but also to make code more generally readable. Just left a few comments - let me know if something comes up or if you think your way should be preferred, but I think it looks good to go once you've made those last few changes!

i agree! all props to kienan for the suggestions
